### ISSUE TITLE
Date range query returns different results in ES7 vs. ES8 - does not respect include_lower

### ISSUE THREAD

**Issue by jubui** [elastic/elasticsearch#109542]
### Elasticsearch Version

8.13.4 (probably all v8.x.y)

### Installed Plugins

none

### Java Version

17

### OS Version

Darwin HH7DY4D3K0 23.5.0 Darwin Kernel Version 23.5.0

### Problem Description

[Date range query returns different results in ES7 vs. ES8 - does not respect include_lower](https://discuss.elastic.co/t/date-range-query-returns-different-results-in-es7-vs-es8-does-not-respect-include-lower/360844). For dates before 1970, in ES8, when using include_lower=false, the date query seems to include the lower bound but in ES7, the date query seems to exclude the lower bound.

### Steps to Reproduce

```
#create index
curl -X PUT localhost:9200/myidx

#create mapping
curl -H "Content-type: application/json" -X PUT localhost:9200/myidx/_mapping -d'{"dynamic":"strict","properties":{"id":{"type":"keyword"},"mydate":{"type":"date","format":"yyyy-MM-ddXXX"}}}'

#Add doc with value 1969-12-31
curl -H "Content-type: application/json" -X POST localhost:9200/myidx/_doc/1 -d'{"id":"1","mydate":"1969-12-31Z"}'

#Query for date range from 1969-12-31 to 2030-10-31 and EXCLUDE the lower and upper endpoints. EXPECT: no result since we are excluding 1969-12-31. The value "-86400000" is 1969-12-31
curl -H "Content-type: application/json" -X POST localhost:9200/myidx/_search -d'{"from":0,"size":5,"query":{"bool":{"must":[{"bool":{"filter":[{"bool":{"must":[{"range":{"mydate":{"from":-86400000,"to":null,"include_lower":false,"include_upper":true,"format":"epoch_millis","boost":1.0}}},{"range":{"mydate":{"from":null,"to":1919635200000,"include_lower":true,"include_upper":false,"format":"epoch_millis","boost":1.0}}}],"adjust_pure_negative":true,"boost":1.0}}],"adjust_pure_negative":true,"boost":1.0}}],"adjust_pure_negative":true,"boost":1.0}},"track_total_hits":2147483647}'
```

Using ES7 via `docker run -e cluster.name=docker-cluster -e cluster.initial_master_nodes=myes -e node.name=myes docker.elastic.co/elasticsearch/elasticsearch:7.17.10` I get 0 results

Using ES8 via `docker run -e xpack.security.enabled=false -e cluster.name=docker-cluster -e cluster.initial_master_nodes=myes -e node.name=myes docker.elastic.co/elasticsearch/elasticsearch:8.13.4` I get 1 result

**Comment by benwtrent** [comment]
This may be due to our Joda changes in v7: https://www.elastic.co/guide/en/elasticsearch/reference/8.14/migrate-to-java-time.html

I am not sure however if it is or not. It strikes me as weird that a new index created in 7. would use Joda unless the format requested specifically required it.

We need to do some digging here to verify.

**Comment by benwtrent** [comment]
If we determine that this was due to the joda change, this issue will be closed as that was a known breaking change between 7 and 8.

**Comment by john-wagster** [comment]
I spent a bit of time looking into this because I had recently been looking at this code.  I believe this is impacting specifically negative epoch millisecond values.  Specifically the way in which the epoch_millis date_parser handles negative values causes a rounding error, which I believe (naively) can be corrected by forcing the parser not to round the value.  I'll dig into a fix to validate this.

This [rounding behavior](https://www.elastic.co/guide/en/elasticsearch/reference/current/query-dsl-range-query.html#missing-date-components) may be unfamiliar but is a result of providing date time values that are not granular enough for comparison, which when providing epoch milliseconds should not be a consideration.

I'll work on a fix for this and report back as soon as I have something.  

@jubui one important side note about 8.x is that `from`, `to`, and `include_lower` has been deprecated and will be removed in 9.x.  The switch to the new api is relatively straight-forward which involves using `gt`, `lt`, `gte` and `lte` instead.  Here's an example below of doing so.  Unfortunately this also exhibits the bug as they share the same underlying date parsing logic.  So there is not a mechanism to mitigate this behavior right now other than what was suggested in the community forum post.

given this mapping
```
  "mappings": {
    "properties": {
      "mydate": {
        "type": "date",
        "format": "yyyy-MM-ddXXX"
      }
    }
  }
```

and this document
```
"mydate": "1969-12-31Z"
```

this query in 7.x 
```
  "query": {
    "range": {
      "mydate": {
        "from": -86400000,
        "include_lower": false
      }
    }
  }
```

would now look like this in 8.x and after (the supported operations are `gt`, `lt`, `gte`, and `lte` and replace the `from` and `to` api such that you no longer need to specify `include_lower`
```
  "query": {
    "range": {
      "mydate": {
        "gt": -86400000
      }
    }
  }
```

**Comment by john-wagster** [comment]
Just to follow up here.  We went back and forth a bit on where this fix needed to live, negative epoch millis have a bit of a history.  

But this PR that was just merged will fix the issue: https://github.com/elastic/elasticsearch/pull/118353

The earliest it will be available is in 8.18. 

Let me know @jubui if you have any additional questions or concerns.  I'lll go ahead and close this issue for now but we can always re-open it or create a new issue.  And thanks for identifying the bug!
